Aliyun: Mock aliyun OSS(Object Storage Service)

xingbowu opened the pull request:
Aliyun OSS is popular in public cloud service, especially the users from china.  Many use cases of integrating iceberg make aliyun oss as backend storage.  So it benefits community to integrate iceberg with aliyun oss. I would like to contribute several PRs and complete this work. Here is 1st step :  Mock aliyun OSS in UT

Aliyun OSS SDK doesn't support mock local  environment,  and no any plan to develop this feature recently.
To make unit test of iceberg integration with oss efficiently, this PR mocks a local lightweight aliyun oss behavior for UT similar with s3mock

xingbowu commented:
@rdblue @openinx Could you help to take a look at this PR? Thanks in advance.

openinx reviewed (commented):
Thanks @xingbowu to make a smaller PR for the Aliyun [OSS integration](https://github.com/apache/iceberg/pull/2230) work.  I skimmed the whole PR,  this PR is trying to introduce the aliyun oss mock app server, so that we could build oss test cases on top of it (We don't need to mock all the called OSS API in a relative complex test cases , such as multi-upload test cases).   It's a good thing for us to introduce a simple simulator to align the local mock app and  aliyun online oss server. 

The most important thing for me is:  How do we guarantee the local mock app is aligning correctly to the aliyun online oss server ?  In the parent PR #2230 ,  we introduced a class test rule named OSSTestRule and the rule has two different impl: 

* OSSMockRule:  The rule will start a local mini aliyun oss server, which could serving the remote OSS client http requests.
* OSSIntegrationTestRule: The rule will prepare testing buckets in the remote oss server, so that the test cases  could write real data to.

For the local oss application,   we provide tests cases in [TestLocalOSS](https://github.com/apache/iceberg/blob/3755c967f61035097abbebfe419d17473bf5fefd/aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/TestLocalOSS.java) , which was designed to run different TestRule according to the intentional environment variables (will run local mock app by default).   If the tests could be passes on both local env and remote oss env, then we could definitely ensure the local oss app is implemented correctly.

xingbowu commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockApp.java:
```diff
@@ -0,0 +1,175 @@
+    @Override
+    public void configureContentNegotiation(final ContentNegotiationConfigurer configurer) {
+      configurer.defaultContentType(MediaType.APPLICATION_FORM_URLENCODED, MediaType.APPLICATION_XML);
+      configurer.favorPathExtension(false);
```
done, remove this one

xingbowu commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockApp.java:
```diff
@@ -0,0 +1,175 @@
+
+  private static class RangeConverter implements Converter<String, Range> {
+
+    private static final Pattern REQUESTED_RANGE_PATTERN = Pattern.compile("^bytes=((\\d*)\\-(\\d*))((,\\d*-\\d*)*)");
```
done

xingbowu commented:
> Thanks @xingbowu to make a smaller PR for the Aliyun [OSS integration](https://github.com/apache/iceberg/pull/2230) work. I skimmed the whole PR, this PR is trying to introduce the aliyun oss mock app server, so that we could build oss test cases on top of it (We don't need to mock all the called OSS API in a relative complex test cases , such as multi-upload test cases). It's a good thing for us to introduce a simple simulator to align the local mock app and aliyun online oss server.
> 
> The most important thing for me is: How do we guarantee the local mock app is aligning correctly to the aliyun online oss server ? In the parent PR #2230 , we introduced a class test rule named OSSTestRule and the rule has two different impl:
> 
> * OSSMockRule:  The rule will start a local mini aliyun oss server, which could serving the remote OSS client http requests.
> * OSSIntegrationTestRule: The rule will prepare testing buckets in the remote oss server, so that the test cases  could write real data to.
> 
> For the local oss application, we provide tests cases in [TestLocalOSS](https://github.com/apache/iceberg/blob/3755c967f61035097abbebfe419d17473bf5fefd/aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/TestLocalOSS.java) , which was designed to run different TestRule according to the intentional environment variables (will run local mock app by default). If the tests could be passes on both local env and remote oss env, then we could definitely ensure the local oss app is implemented correctly.



> Thanks @xingbowu to make a smaller PR for the Aliyun [OSS integration](https://github.com/apache/iceberg/pull/2230) work. I skimmed the whole PR, this PR is trying to introduce the aliyun oss mock app server, so that we could build oss test cases on top of it (We don't need to mock all the called OSS API in a relative complex test cases , such as multi-upload test cases). It's a good thing for us to introduce a simple simulator to align the local mock app and aliyun online oss server.
> 
> The most important thing for me is: How do we guarantee the local mock app is aligning correctly to the aliyun online oss server ? In the parent PR #2230 , we introduced a class test rule named OSSTestRule and the rule has two different impl:
> 
> * OSSMockRule:  The rule will start a local mini aliyun oss server, which could serving the remote OSS client http requests.
> * OSSIntegrationTestRule: The rule will prepare testing buckets in the remote oss server, so that the test cases  could write real data to.
> 
> For the local oss application, we provide tests cases in [TestLocalOSS](https://github.com/apache/iceberg/blob/3755c967f61035097abbebfe419d17473bf5fefd/aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/TestLocalOSS.java) , which was designed to run different TestRule according to the intentional environment variables (will run local mock app by default). If the tests could be passes on both local env and remote oss env, then we could definitely ensure the local oss app is implemented correctly.

Thanks a lot for pointing out missing part. I have done rework and added local test here to guarantee quality. comparing with 2230,  implemented more test case, such as range get. 

Additionally, followed up your review comments to simulate basic aliyun oss behavior including create/put/delete and excluded multi-parts , feel free to let me know if you have further comments.

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/util/AliyunOSSURI.java:
```diff
@@ -0,0 +1,86 @@
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+
+public class AliyunOSSURI {
```
Do we really need to introduce this class in this PR ?  I see no place are referring to this class.

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/AliyunOSSTestRule.java:
```diff
@@ -0,0 +1,84 @@
+public interface AliyunOSSTestRule extends TestRule {
+  Logger LOG = LoggerFactory.getLogger(AliyunOSSTestRule.class);
+  UUID RANDOM_UUID = java.util.UUID.randomUUID();
+  /**
```
In apache iceberg project, we usually leave an empty line before the two code blocks between two methods declaration.  I guess you need to check your intellij code style configuration and also apply the iceberg [checkstyles](https://iceberg.apache.org/community/#setting-up-ide-and-code-style) from the document. 

![image](https://user-images.githubusercontent.com/5028729/132298216-ab4cbdb3-43d1-4bad-bbe9-4e5d148be100.png)

xingbowu replied:
I have checked the idea style file "intellij-java-palantir-style.xml",  the default style between two method in the interface is empty as below 
![image](https://user-images.githubusercontent.com/12458911/133087831-216e217b-46a2-4915-b7c5-66e3468d6905.png)
let me know if the style file is incorrect

openinx replied:
I think it's worth to publish a PR to correct the option value in the [intellij-java-palantir-style.xml](https://github.com/apache/iceberg/blob/master/.baseline/idea/intellij-java-palantir-style.xml#L454).

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/AliyunOSSTestRule.java:
```diff
@@ -0,0 +1,84 @@
+import org.junit.runners.model.Statement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```
I think we need to add javadoc describing why we need to introduce the `AliyunOSSTestRule` ?

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/AliyunOSSTestRule.java:
```diff
@@ -0,0 +1,84 @@
+import org.slf4j.LoggerFactory;
+
+public interface AliyunOSSTestRule extends TestRule {
+  Logger LOG = LoggerFactory.getLogger(AliyunOSSTestRule.class);
```
This line can be removed now ? Seems nobody are referring this.

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockRule.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    public AliyunOSSMockRule build() {
+      if (props.get(AliyunOSSMockApp.PROP_ROOT_DIR) == null) {
```
So when the PROP_ROOT_DIR is not set or set to empty string,  we will use the default temporary directory with a current timestamp suffix.  Then we will create the directory by using the `createRootDir` method .  I think it's more clear to decide which root dir value we will use firstly, then check whether it's created for testing: 

```java
    public AliyunOSSMockRule build() {
      String rootDir = (String) props.get(AliyunOSSMockApp.PROP_ROOT_DIR);
      if (StringUtils.isNullOrEmpty(rootDir)) {
        File dir = new File(FileUtils.getTempDirectory(), "oss-mock-file-store" + System.currentTimeMillis());
        rootDir = dir.getAbsolutePath();
      }
      File root = new File(rootDir);
      root.deleteOnExit();
      root.mkdir();

      return new AliyunOSSMockRule(props);
    }
```

openinx replied:
Maybe we will still need to set the rootDir back to the properties Map<String, Object> in the above code.  I'm trying not to access the PROP_ROOT_DIR again and again in the current PR.

xingbowu replied:
yes, we need set the rootDir only if no such element in the properties Map.

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockRule.java:
```diff
@@ -0,0 +1,143 @@
+            try {
+              Files.delete(p);
+            } catch (IOException e) {
+              // delete this files quietly.
```
Nit: `files` - > `file`

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockRule.java:
```diff
@@ -0,0 +1,143 @@
+  public void tearDownBucket(String bucket) {
+    try {
+      Files.walk(rootDir().toPath())
+          .filter(p -> p.toFile().isFile())
```
Why if we have a bucket named `test-bucket`, and have a object named `path/to/a.dat`,  Then the full local path will be `<root-dir>/test-bucket/path/to/a.dat`,   will we remove the `<root-dir>/test-bucket/path/to` directory also ?

openinx replied:
In aliyun OSS,  the `<root-dir>/test-bucket/path/to` won't be a specific directory in the real production, but in our local OSS storage app,  it will be a real directory (though it's not a real object that people could see by using aliyun OSS SDK).

xingbowu replied:
Regarding to the scenario you list above,  the directories are deleted in the deleteBucket. However,  we indeed have problem in another scenario. if we create an object with name "aa/bb/cc.txt", and then remove it. after that we create a new object with name "aa/bb". it has problem because of deleting cc.txt in current logic only. I will fix it in separate PR.

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockRule.java:
```diff
@@ -0,0 +1,143 @@
+  public static class Builder {
+    private Map<String, Object> props = Maps.newHashMap();
+
+    public Builder withRootDir(String rootDir) {
```
I'm thinking we may don't need to expose a public `withRootDir` in this Builder because people often don't set any specific RootDir in their mocking unit tests, they just use the temporary directory directly.

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockApp.java:
```diff
@@ -0,0 +1,160 @@
+  @Autowired
+  private ConfigurableApplicationContext context;
+
+  public static void main(final String[] args) {
+    start(Maps.newHashMap(), args);
+  }
```
Is their any place that we will start a main thread to start the application in iceberg unit tests ?  If don't, then we iceberg usually don't expose any `main` thread.

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/TestLocalAliyunOSS.java:
```diff
@@ -0,0 +1,184 @@
+public class TestLocalAliyunOSS {
+
+  @ClassRule
+  public static final AliyunOSSTestRule OSS_TEST_RULE = AliyunOSSMockRule.builder().silent().build();
```
I think we need to make this `TestRule` to be configurable so that we could verify this unit tests on both local mock oss services and remote aliyun OSS environment, to guarantee that the local oss app has the same semantics as remote aliyun OSS environment.

openinx replied:
@xingbowu How is feeling about this comment ?

xingbowu replied:
yes, I will implement it in the next PR, which include both local and remote part

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/TestLocalAliyunOSS.java:
```diff
@@ -0,0 +1,184 @@
+
+  private boolean doesBucketExist(String bucket) {
+    try {
+      oss.createBucket(bucket);
```
Em.. seems we don't provide any bucket existence checking API in the local OSS.  I forgot why we don't implement it ,  seems it's fussy to mock all the object metadata..

xingbowu replied:
it is the simplest way to simulate this api behavior.  we can mock further version if necessary later per upper layer need. may keep it as it is here.

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/TestLocalAliyunOSS.java:
```diff
@@ -0,0 +1,184 @@
+
+  @Test
+  public void testDoesObjectExist() {
+    Assert.assertFalse(oss.doesObjectExist(bucketName, "key"));
+
```
Why we need two same assert sentences ?

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/TestLocalAliyunOSS.java:
```diff
@@ -0,0 +1,184 @@
+
+  @Test
+  public void testGetObjectWithRange() throws IOException {
+
```
Extra empty line ?

openinx replied:
Nit: I think this line should be removed because it seems not match the whole iceberg project code style , even if we don't put it into the check style rule set.

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/TestLocalAliyunOSS.java:
```diff
@@ -0,0 +1,184 @@
+    getObjectRequest.setRange(start, end);
+    IOUtils.readFully(oss.getObject(getObjectRequest).getObjectContent(), actual);
+
+    Assert.assertArrayEquals(bytes, actual);
```
It's incorrect to test the object range query by asserting the whole 2000 bytes to the bytes which is read from range [0, 1999], because it did not cover the case that `range` does not affect anything.  If we want to design case to cover the range read, then we must cover the following cases: 

* Range = [0,0]
* Range = [0,1]
* Range = [1, 100]
* Range = [0, 1999]
* Range = [-1, 100]
* Range = [100, -1]
* Range = [-1, -1]

etc.

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockLocalStore.java:
```diff
@@ -0,0 +1,197 @@
+        buckets.add(new Bucket(path.getFileName().toString()));
+      }
+    } catch (final IOException e) {
+      LOG.error("Could not Iterate over Bucket-Folders", e);
```
Nit: `Iterate` -> `iterate`

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockLocalStore.java:
```diff
@@ -0,0 +1,197 @@
+
+  Bucket getBucket(String bucketName) {
+    List<Bucket> buckets = findBucketsByFilter(file ->
+        Files.isDirectory(file) && file.getFileName().endsWith(bucketName));
```
> file.getFileName().endsWith(bucketName)

It's incorrect to use the `file.getFileName().endsWith(bucketName)` to check its bucket name, because for a bucket `test-bucket`, it's possible that we will have an object with name `/path/to/test-bucket/a.txt`.  In that case,  we will create a directory `<root-dir>/path/to/test-bucket` though it's no an object being visiable to OSS SDK, but it will fail to check the existence of this given bucket.

xingbowu replied:
newDirectoryStream only traverses the directory&file under root path. the object with name /test-bucket/a.txt, is under <root-dir>/test-bucket(bucket name)/path/to/test-bucket(object prefix name)/a.txt. so it is no problem here.

openinx reviewed (changes requested):
@xingbowu Thanks for the updating,  I just token another round  and left some comments !

openinx commented:
The broken unit test is : 

```
org.apache.iceberg.types.TestReadabilityChecks > testStructWriteReordering STANDARD_ERROR
    [nested.field_b is out of order, before field_a]

org.apache.iceberg.io.TestCloseableGroup > suppressExceptionIfSetSuppressIsTrue STANDARD_ERROR
    [Test worker] ERROR org.apache.iceberg.io.CloseableGroup - Exception suppressed when attempting to close resources
    java.io.IOException: exception1
    	at org.apache.iceberg.io.CloseableGroup.close(CloseableGroup.java:80)
    	at org.apache.iceberg.io.TestCloseableGroup.suppressExceptionIfSetSuppressIsTrue(TestCloseableGroup.java:75)
    	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
    	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
    	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
    	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
    	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
    	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
    	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
    	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
    	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
    	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
    	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
    	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
    	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
    	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
    	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
    	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
    	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
    	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
    	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
    	at jdk.internal.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
    	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
    	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
    	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
    	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
    	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:118)
    	at jdk.internal.reflect.GeneratedMethodAccessor3.invoke(Unknown Source)
    	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
    	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:175)
    	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:157)
    	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
    	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
    	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
    	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
    	at java.base/java.lang.Thread.run(Thread.java:829)
```

Looks like it's unrelated to this PR,  let's reopen to trigger the travis CI once again.

xingbowu commented:
@openinx  Thanks for your effort and comments, I have reworked them with latest code, feel free to take further review

openinx commented on build.gradle:
```diff
@@ -262,6 +262,36 @@ project(':iceberg-data') {
+    compile project(':iceberg-common')
+
+    compileOnly 'com.aliyun.oss:aliyun-sdk-oss'
+    compileOnly 'javax.xml.bind:jaxb-api'
+    compileOnly 'javax.activation:activation'
+    compileOnly 'org.glassfish.jaxb:jaxb-runtime'
```
Why do we need to introduce those three jars when adding a testing framework under the aliyun/src/test module?

xingbowu replied:
per aliyun oss sdk documentation, https://help.aliyun.com/document_detail/32009.html
jaxb related dependencies need to be added under java 9 and plus version environment.

openinx replied:
Okay, then I will recommend to package all the dependencies jar into a single bundled iceberg-aliyun jar. Because in flink SQL client,  we will need to add [each jar](https://iceberg.apache.org/flink/#preparation-when-using-flink-sql-client) in the shell as the following: 

```shell
./bin/sql-client.sh embedded \
-j <iceberg-aliyu>.jar \
-j <aliyun-oss>.jar \
-j <javax.xml.bind:jaxb-api>.jar \
-j <org.glassfish.jaxb:jaxb-runtime>.jar 
shell
```

It will be quite tedious for people to add jars one by one to make the iceberg job works ( aws-sdk don't need to package all of them into a bundled jar because aws sdk has provided [it](https://mvnrepository.com/artifact/software.amazon.awssdk/bundle) ( see [document](https://iceberg.apache.org/aws/#enabling-aws-integration)).

openinx commented on versions.props:
```diff
@@ -16,6 +16,10 @@ com.github.ben-manes.caffeine:caffeine = 2.8.4
 org.apache.arrow:arrow-memory-netty = 2.0.0
 com.github.stephenc.findbugs:findbugs-annotations = 1.3.9-1
+com.aliyun.oss:aliyun-sdk-oss = 3.10.2
+javax.xml.bind:jaxb-api = 2.3.1
+javax.activation:activation = 1.1.1
+org.glassfish.jaxb:jaxb-runtime = 2.3.3
```
Why we need to three extra jars  ?  I passed the test cases after removing them ..

xingbowu replied:
same as above comment

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/AliyunOSSTestRule.java:
```diff
@@ -0,0 +1,82 @@
+import org.junit.runners.model.Statement;
+
+/**
+ * API for test Aliyun Object Storage Service (OSS) either local mock or remote server
```
> API for test Aliyun Object Storage Service (OSS) either local mock or remote server

Change it to `API for test Aliyun Object Storage Service (OSS)  which is either local mock http server or remote aliyun oss server`

openinx commented:
I left several comments, almost looks good to me.  @rdblue @jackye1995 would you like to have a double check ?

openinx commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockApp.java:
```diff
@@ -0,0 +1,156 @@
+
+    Banner.Mode bannerMode = Banner.Mode.CONSOLE;
+
+    if (Boolean.parseBoolean(String.valueOf(properties.remove("silent")))) {
```
Nit:  Here we can replace the `silent` by the `PROP_SILENT`.

openinx commented:
@xingbowu ,  Could you please file new issues for the above comment ([1], [2], [3]) and put them in this project dashboard , so that we could easily track the whole progress  ? 

I plan to get this PR merged once we've tracked those issues and fixed the other minor comments, thanks for the work !

[1] https://github.com/apache/iceberg/pull/3067#discussion_r703249976
[2] https://github.com/apache/iceberg/pull/3067#discussion_r703261388
[3] https://github.com/apache/iceberg/pull/3067#discussion_r711561677

xingbowu commented:
> @xingbowu , Could you please file new issues for the above comment ([1], [2], [3]) and put them in this project dashboard , so that we could easily track the whole progress ?
> 
> I plan to get this PR merged once we've tracked those issues and fixed the other minor comments, thanks for the work !
> 
> [1] [#3067 (comment)](https://github.com/apache/iceberg/pull/3067#discussion_r703249976)
> [2] [#3067 (comment)](https://github.com/apache/iceberg/pull/3067#discussion_r703261388)
> [3] [#3067 (comment)](https://github.com/apache/iceberg/pull/3067#discussion_r711561677)

Issue #3180 opened

openinx reviewed (approved):
LGTM !
